velox/exec/PartitionedOutputBufferManager.h

```diff
@@ -40,22 +40,62 @@ struct DataAvailable {
   }
 };
 
-class DestinationBuffer {
+/// The class is used to buffer the output pages which haven't been fetched by
+/// any destination for arbitrary output.
+///
+/// NOTE: this class is not thread-safe.
+class ArbitraryBuffer {
  public:
-  void enqueue(std::shared_ptr<SerializedPage> data) {
-    // drop duplicate end markers
-    if (data == nullptr && !data_.empty() && data_.back() == nullptr) {
-      return;
-    }
+  ArbitraryBuffer() = default;
+
+  /// Returns true if this arbitrary buffer has no buffered pages.
+  bool empty() const {
+    return pages_.empty() || hasNoMoreData();
+  }
 
-    data_.push_back(std::move(data));
+  /// Returns true if this shared buffer has no more incoming pages but it can
+  /// still has buffered pages.
+  bool hasNoMoreData() const {
+    return !pages_.empty() && (pages_.back() == nullptr);
   }
 
+  void enqueue(std::unique_ptr<SerializedPage> page);
+
+  /// Gets a number of pages with total bytes up to 'maxBytes'.
+  std::vector<std::shared_ptr<SerializedPage>> getPages(uint64_t maxBytes);
+
+  std::string toString() const;
+
+ private:
+  std::deque<std::shared_ptr<SerializedPage>> pages_;
+};
+
+class DestinationBuffer {
+ public:
+  DestinationBuffer() = default;
+
+  void enqueue(std::shared_ptr<SerializedPage> data);
+
+  /// Invoked to load data with up to 'notifyMaxBytes_' bytes from arbitrary
+  /// 'buffer' if there is pending fetch from this destination in which case
+  /// 'notify_' is not null. Otherwise, it does nothing. This only used by
+  /// arbitrary output type when enqueue new data.
+  void maybeLoadData(ArbitraryBuffer* buffer);
+
+  /// Invoked to load data with up to 'maxBytes' from arbitrary 'buffer' when
+  /// fetch data from this destination. This only used by arbitrary output type
+  /// which doesn't expect to buffer any data and is always load data from the
+  /// arbitrary buffer on demand.
+  void loadData(ArbitraryBuffer* buffer, uint64_t maxBytes);
+
   // Returns a shallow copy (folly::IOBuf::clone) of the data starting at
   // 'sequence', stopping after exceeding 'maxBytes'. If there is no data,
   // 'notify' is installed so that this gets called when data is added.
-  std::vector<std::unique_ptr<folly::IOBuf>>
-  getData(uint64_t maxBytes, int64_t sequence, DataAvailableCallback notify);
+  std::vector<std::unique_ptr<folly::IOBuf>> getData(
+      uint64_t maxBytes,
+      int64_t sequence,
+      DataAvailableCallback notify,
```

Comment:
Would it make sense to refactor this class to use Futures?

Reply:
Not sure what do you want to refactor here. Let's chat offline about this for a followup. Thanks!